File: androidApp/src/main/java/com/ramo/workoutly/android/global/base/Theme.kt
package com.ramo.workoutly.android.global.base

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Shapes
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

data class Theme(
    val isDarkMode: Boolean,
    val isDarkStatusBarText: Boolean,
    val primary: Color,
    val secondary: Color,
    val background: Color,
    val backgroundGradient: Brush,
    val statusBarColor: Color,
    val backDark: Color,
    val backDarkSec: Color,
    val backgroundPrimary: Color,
    val backGreyTrans: Color,
    val textColor: Color,
    val textForPrimaryColor: Color,
    val textGrayColor: Color,
    val error: Color,
    val textHintColor: Color,
    val pri: Color
) {
    @Suppress("unused")
    val priAlpha: Color = pri.copy(alpha = 0.14F)
    val textHintAlpha = textHintColor.copy(alpha = 0.5F)
    val backDarkAlpha = backDark.copy(alpha = 0.5F)
    val seenColor = if (isDarkMode) Color(0xFF4FC3F7) else Color.Blue
}

fun generateTheme(isDarkMode: Boolean): Theme {
    return if (isDarkMode) {
        Theme(
            isDarkMode = true,
            isDarkStatusBarText = false,
            primary = Color(red = 255, green = 74, blue = 74),//rgb(255, 74, 74)
            secondary = Color(red = 231, green = 83, blue = 83, alpha = 255),
            background = Color(red = 31, green = 31, blue = 31, alpha = 255),
            backgroundGradient = Color(0xFF265160).let { it.getGradientFromBaseColor(it.darken(0.2F)) },
            statusBarColor = Color(0xFF265160).darken(0.2F),
            backDark = Color(50, 50, 50),
            backDarkSec = Color(100, 100, 100),
            backgroundPrimary = Color(red = 31, green = 31, blue = 31, alpha = 255).darker(),
            backGreyTrans = Color(85, 85, 85, 85),
            textColor = Color.White,
            textForPrimaryColor = Color(red = 204, green = 204, blue = 204),
            textGrayColor = Color(143, 143, 143),
            error = Color(red = 255, green = 21, blue = 21),
            textHintColor = Color(red = 204, green = 204, blue = 204),
            pri = Color(102, 158, 255)
        )
    } else {
        Theme(
            isDarkMode = false,
            isDarkStatusBarText = true,
            primary = Color(red = 255, green = 74, blue = 74),//rgb(255, 74, 74)
            secondary = Color(red = 231, green = 83, blue = 83, alpha = 255),
            background = Color.White,
            backgroundGradient = Color(0xFF265160).let { it.getGradientFromBaseColor(it.lighten(0.2F)) },
            statusBarColor = Color(0xFF265160).lighten(0.2F),
            backDark = Color(215, 215, 215),
            backDarkSec = Color(200, 200, 200),
            backgroundPrimary = Color.White.darker(),
            backGreyTrans = Color(170, 170, 170, 85),
            textColor = Color.Black,
            textForPrimaryColor = Color.White,
            textGrayColor = Color(112, 112, 112),
            error = Color(red = 155, green = 0, blue = 0),
            textHintColor = Color(red = 68, green = 68, blue = 68),
            pri = Color(102, 158, 255)
        )
    }
}


@Composable
fun MyApplicationTheme(
    theme: Theme,
    content: @Composable () -> Unit
) {
    val colors = if (isSystemInDarkTheme()) {
        darkColorScheme(
            primary = theme.primary,
            secondary = theme.secondary,
            background = theme.background,
            onBackground = theme.textHintColor,
        )
    } else {
        lightColorScheme(
            primary = theme.primary,
            secondary = theme.secondary,
            background = theme.background,
            onBackground = theme.textHintColor,
        )
    }
    val typography = Typography(
        bodyMedium = TextStyle(
            fontFamily = FontFamily.Default,
            fontWeight = FontWeight.Normal,
            fontSize = 16.sp
        )
    )
    val shapes = Shapes(
        small = RoundedCornerShape(4.dp),
        medium = RoundedCornerShape(4.dp),
        large = RoundedCornerShape(0.dp)
    )
    MaterialTheme(
        colorScheme = colors,
        typography = typography,
        shapes = shapes,
        content = content
    )
}

File: iosApp/iosApp/global/base/Theme.swift
import SwiftUI

struct Theme {

    let isDarkMode: Bool
    let isDarkStatusBarText: Bool
    let primary: Color
    let secondary: Color
    let background: Color
    let surface: Color
    let backDark: Color
    let backDarkSec: Color
    let backGreyTrans: Color
    let textColor: Color
    let textForPrimaryColor: Color
    let textGrayColor: Color
    let error: Color
    let textHintColor: Color
    let textHintAlpha: Color
    let backDarkAlpha: Color
    let primaryAlpha: Color
    let backgroundPrimary: Color
    let pri: Color
    let priAlpha: Color
    let seenColor: Color
    
    init(isDarkMode: Bool) {
        self.isDarkMode = isDarkMode
        if (isDarkMode) {
            self.isDarkStatusBarText = false
            self.primary = Color(red: 1, green: 74 / 255, blue: 74 / 255)
            self.secondary = Color(red: 231 / 255, green: 83 / 255, blue: 83 / 255)
            self.surface =  Color(red: 31 / 255, green: 31 / 255, blue: 31 / 255)
            self.backgroundPrimary = Color(red: 31 / 255, green: 31 / 255, blue: 31 / 255).margeWithPrimary
            self.background = Color(red: 31 / 255, green: 31 / 255, blue: 31 / 255)
            self.backDark = Color(red: 50 / 255, green: 50 / 255, blue: 50 / 255)
            self.backDarkSec = Color(red: 100 / 255, green: 100 / 255, blue: 100 / 255)
            self.backGreyTrans = UIColor(red: 85 / 255, green: 85 / 255, blue: 85 / 255, alpha: 85 / 255).toC
            self.textColor = Color.white
            self.textForPrimaryColor = Color.black
            self.textGrayColor = Color(red: 143 / 255, green: 143 / 255, blue: 143 / 255)
            self.error = Color(red: 1, green: 21 / 255, blue: 21 / 255)
            self.textHintColor = Color(red: 204 / 255, green: 204 / 255, blue: 204 / 255) // Color.LightGray
            self.pri = Color(red: 102 / 255, green: 158 / 255, blue: 1)
            self.textHintAlpha = UIColor(red: 175 / 255, green: 175 / 255, blue: 175 / 255, alpha: 0.5).toC
            self.backDarkAlpha = UIColor(red: 31 / 255, green: 31 / 255, blue: 31 / 255, alpha: 0.5).toC
            self.primaryAlpha = UIColor(red: 208 / 255, green: 188 / 255, blue: 1, alpha: 0.5).toC
            self.priAlpha = UIColor(red: 109 / 255, green: 157 / 255, blue: 241 / 255, alpha: 0.14).toC
            self.seenColor = Color(red: 79 / 255, green: 195 / 255, blue: 247 / 255)
        } else {
            self.isDarkStatusBarText = true
            self.primary = Color(red: 1, green: 74 / 255, blue: 74 / 255)
            self.secondary = Color(red: 231 / 255, green: 83 / 255, blue: 83 / 255)
            self.surface =  Color.white
            self.backgroundPrimary = Color.white.margeWithPrimary
            self.background = Color.white
            self.backDark = Color(red: 215 / 255, green: 215 / 255, blue: 215 / 255)
            self.backDarkSec = Color(red: 200 / 255, green: 200 / 255, blue: 200 / 255)
            self.backGreyTrans = UIColor(red: 170 / 255, green: 170 / 255, blue: 170 / 255, alpha: 85 / 255).toC
            self.textColor = Color.black
            self.textForPrimaryColor = Color.black
            self.textGrayColor = Color(red: 112 / 255, green: 112 / 255, blue: 112 / 255)
            self.error = Color(red: 155 / 255, green: 0, blue: 0)
            self.textHintColor = Color(red: 68 / 255, green: 68 / 255, blue: 68 / 255) // Color.DarkGray
            self.pri = Color(red: 102 / 255, green: 158 / 255, blue: 1)
            self.textHintAlpha = UIColor(red: 80 / 255, green: 80 / 255, blue: 80 / 255, alpha: 0.5).toC
            self.backDarkAlpha = UIColor(red: 241 / 255, green: 241 / 255, blue: 241 / 255, alpha: 0.5).toC
            self.primaryAlpha = UIColor(red: 102 / 255, green: 80 / 255, blue: 164 / 255, alpha: 0.5).toC
            self.priAlpha = UIColor(red: 1, green: 1, blue: 1, alpha: 0.14).toC
            self.seenColor = Color.blue
        }
    }
    
    /*func colorScheme(screen: Screen) -> ColorScheme {
        return screen == .SPLASH_SCREEN_ROUTE ? (isDarkMode ? ColorScheme.dark : ColorScheme.light) : (isDarkStatusBarText ? ColorScheme.light : ColorScheme.dark)
    }*/
    
    func textFieldColor(isError: Bool, isEmpty: Bool) -> Color {
        if (isError) {
            return error
        } else {
            return isEmpty ? Color.black.opacity(0.7) : Color.black
        }
    }
}
